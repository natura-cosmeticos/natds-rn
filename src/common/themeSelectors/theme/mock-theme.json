{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
