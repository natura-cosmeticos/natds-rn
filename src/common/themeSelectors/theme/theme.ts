import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
