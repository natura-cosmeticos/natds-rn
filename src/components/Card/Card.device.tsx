import { storiesOf } from '@storybook/react-native'
import { Default, Interactive } from './Card.stories'

storiesOf('Card', module)
  .add('Default', Default)
  .add('Interactive', Interactive)
