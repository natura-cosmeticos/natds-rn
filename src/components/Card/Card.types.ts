import { ReactNode } from 'react'

export type CardTypes = 'base';

export interface CardProps {
  /**
   * The type of card to render.
   */
  type?: CardTypes;
  /**
   * The Card content
   */
  children: ReactNode;
  /**
   * Optional ID for testing
   */
  testID?: string;
  /**
   * Render a elevation shadow on the card
   * @default true
   */
  elevation?: boolean;
  /**
   * Determine rouded corners in the Card
   * @default true
   */
  radius?: boolean;
}
