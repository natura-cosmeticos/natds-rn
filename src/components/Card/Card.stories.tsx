import React from 'react'
import { View } from 'react-native'
import { select, text as textKnob } from '@storybook/addon-knobs'
import { Card, CardTypes } from './Card'
import { TextWithTheme } from '../../common/HelperComponents/ThemeHelper.styles'

export default {
  component: Card,
  parameters: {
    componentSubtitle:
      'Cards are surfaces that groups content and actions in the same space. Its construction must always create a visual hierarchy of the content and communicate its purposes clearly.'
  },
  title: 'Components/Card'
}

const cardTypes = {
  base: 'base'
}

export const base = () => (
  <View
    style={{
      alignContent: 'center',
      justifyContent: 'center',
      margin: 32,
      width: 328
    }}
  >
    <TextWithTheme style={{ marginBottom: 8 }}>Base</TextWithTheme>
    <Card>
      <View style={{ minHeight: 192 }} />
    </Card>
  </View>
)

export const interactive = () => (
  <View
    style={{
      alignContent: 'center',
      justifyContent: 'center',
      margin: 32,
      width: 328
    }}
  >
    <TextWithTheme style={{ marginBottom: 8 }}>Interactive</TextWithTheme>
    <Card type={select('Types', cardTypes, 'base') as CardTypes}>
      <View style={{ minHeight: 192 }}>
        <TextWithTheme style={{ margin: 16 }}>
          {textKnob('Content', 'Some content here.')}
        </TextWithTheme>
      </View>
    </Card>
  </View>
)
