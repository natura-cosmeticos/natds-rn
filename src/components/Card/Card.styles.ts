import styled from 'styled-components/native'
import { css } from 'styled-components'

import {
  Theme,
  getColorSurface,
  getShadowBySize
} from '../../common/themeSelectors'
import { CardProps } from './Card.types'

type CardStyleProps = {
  theme: Theme;
} & Pick<CardProps, 'radius' | 'elevation'>

export const CardBase = styled.View`
  ${({ theme, radius, elevation = true }: CardStyleProps) => css`
    ${elevation ? getShadowBySize(theme, 'micro') : { elevation: 0 }})}
    background-color: ${getColorSurface(theme)};
    border-radius: ${radius ? `${theme.card.borderRadius.enable}px` : `${theme.card.borderRadius.disable}px`};
  `}
`
