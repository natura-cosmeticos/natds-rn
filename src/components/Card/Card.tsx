import React from 'react'

import { CardProps } from './Card.types'
import { CardBase } from './Card.styles'

export const Card = ({
  children,
  testID = 'card',
  type = 'base',
  elevation = true,
  radius = true
}: CardProps) => (
  <CardBase
    testID={testID}
    type={type}
    elevation={elevation}
    radius={radius}
  >
    {children}
  </CardBase>
)
