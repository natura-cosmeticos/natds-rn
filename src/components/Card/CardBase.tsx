import React from 'react'

import { withTheme } from 'styled-components/native'
import { Theme, getShadowBySize } from '../../common/themeSelectors'
import { CardProps } from './Card.types'
import { CardBase } from './Card.styles'

type CardBaseProps = CardProps & { theme: Theme }

export const CardComponent = ({
  children,
  testID,
  type,
  elevation,
  radius,
  theme
}: CardBaseProps) => (
  <CardBase
    style={elevation && getShadowBySize(theme, 'micro')}
    testID={testID}
    type={type}
    elevation={elevation}
    radius={radius}
  >
    {children}
  </CardBase>
)

export const CardBaseComponent = withTheme(CardComponent)
