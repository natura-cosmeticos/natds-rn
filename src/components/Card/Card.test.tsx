import '@testing-library/jest-native/extend-expect';
import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from 'styled-components/native';
import { Text } from 'react-native';
import theme from '../../common/themeSelectors/theme/mock-theme.json';
import { Card, CardTypes } from './Card';

jest.useFakeTimers();

const renderCard = (fn, props?: CardTypes) => fn(
    <ThemeProvider theme={theme}>
      <Card>
        <Text>{props}</Text>
      </Card>
    </ThemeProvider>,
);

describe('Card component', () => {
  it('should render Base type as default', () => {
    const { queryByTestId } = renderCard(render);

    expect(queryByTestId('card')?.props).toHaveProperty('type', 'base');
  });
  it('should render Card component', () => {
    const card = renderCard(renderer.create);

    expect(card).toMatchSnapshot();
  });
});
