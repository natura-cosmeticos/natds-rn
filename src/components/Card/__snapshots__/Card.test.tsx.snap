// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render a Card without elevation 1`] = `
<View
  elevation={false}
  radius={true}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 4,
        "elevation": 0,
      },
    ]
  }
  testID="elevationID"
  type="base"
>
  <Text>
    Card
  </Text>
</View>
`;

exports[`Card component should render a Card without rounded corners 1`] = `
<View
  elevation={true}
  radius={false}
  style={
    Array [
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 0,
        "elevation": 1,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 1,
          "width": 0,
        },
        "shadowOpacity": 0.14,
        "shadowRadius": 1,
      },
    ]
  }
  testID="radiusID"
  type="base"
>
  <Text>
    Card
  </Text>
</View>
`;
